core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -739,56 +738,57 @@ protected File getRootDirFor(String name) {
 
     /**
      * Gets the {@link Jenkins} singleton.
-     * {@link #getInstanceOrNull()} provides the unchecked versions of the method.
      * @return {@link Jenkins} instance
-     * @throws IllegalStateException {@link Jenkins} has not been started, or was already shut down
-     * @since 1.590
-     * @deprecated use {@link #getInstance()}
+     * @throws IllegalStateException for the reasons that {@link #getInstanceOrNull} might return null
+     * @since FIXME
      */
-    @Deprecated
     @Nonnull
-    public static Jenkins getActiveInstance() throws IllegalStateException {
-        Jenkins instance = HOLDER.getInstance();
+    public static Jenkins get() throws IllegalStateException {
+        Jenkins instance = getInstanceOrNull();
         if (instance == null) {
-            throw new IllegalStateException("Jenkins has not been started, or was already shut down");
+            throw new IllegalStateException("Jenkins.instance is missing. Read the documentation of Jenkins.getInstanceOrNull to see what you are doing wrong.");
         }
         return instance;
     }
 
+    /**
+     * @deprecated This is a verbose historical alias for {@link #get}.
+     * @since 1.590
+     */
+    @Deprecated
+    @Nonnull
+    public static Jenkins getActiveInstance() throws IllegalStateException {
+        return get();
+    }
+
     /**
      * Gets the {@link Jenkins} singleton.
-     * {@link #getActiveInstance()} provides the checked versions of the method.
-     * @return The instance. Null if the {@link Jenkins} instance has not been started,
-     * or was already shut down
+     * {@link #get} is what you normally want.
+     * <p>In certain rare cases you may have code that is intended to run before Jenkins starts or while Jenkins is being shut down.
+     * For those rare cases use this method.
+     * <p>In other cases you may have code that might end up running on a remote JVM and not on the Jenkins master.
+     * For those cases you really should rewrite your code so that when the {@link Callable} is sent over the remoting channel
+     * it can do whatever it needs without ever referring to {@link Jenkins};
+     * for example, gather any information you need on the master side before constructing the callable.
+     * If you must do a runtime check whether you are in the master or agent, use {@link JenkinsJVM} rather than this method,
+     * as merely loading the {@link Jenkins} class file into an agent JVM can cause linkage errors under some conditions.
+     * @return The instance. Null if the {@link Jenkins} service has not been started, or was already shut down,
+     *         or we are running on an unrelated JVM, typically an agent.
```

Comment:
I do wonder why this doesn't check that it is running on the master and throw an exception if not (as that is a clear violation of the intent of this contract) and from this java client something that should never be done.

Reply:
I agree a new method for this should have a 

```java
if (!JenkinsJVM.isJenkinsJVM()) {
    throw new IllegalStateException("Jenkins instance must not be accessed outside Jenkins");
}
````

by possibly deprecating it and creating a `getOrNull()` to be consistent with the new methods naming. (?)

Reply:
Does not work I think—`isJenkinsJVM` will return false for the same reasons that `Jenkins.instance == null`. It just avoids loading `Jenkins.class` (and thus `javax.servlet.*`, …) in the process.

Reply:
Not sure what you mean, but the goal would be to have an exception thrown instead of just returning `null` so that plugins developers don't put an `if` on the nullity of `getInstanceOrNull()` anymore when this is always going to be null on an agent's side?

Reply:
Not sure where you are going with this. What use case?

Normal code can simply use `.get()` and be done.

Reply:
@batmat well at some point we will actually ban `Jenkins` from the agent classloader, so anyone doing that will just plain blow up there and then. I don't think we need to guard further as in that case you still would not get the exception, just a class not found